contributors/design-proposals/secure_watch.md

```diff
@@ -0,0 +1,208 @@
+being referenced by this pods.
+2. From that point, events for all objects being referenced by this pod (e.g.
+add/update/delete of secret being referenced by this pod) are being delivered
+to the watcher as long as the pod exists on the node.
+
+ With this semantic we will be able to reuse e.g. reflector framework with just
+minor modifications.
+
+TODO: Describe how this should be used.
+
+
+### Determining whether object is referenced by pods from a given node.
+
+ The tricky part in determining whether an object X is referenced by any pods
+bound to a given node is to avoid different kind of race conditions and do it
+in deterministic way.
+
+ The crucial requirements here are:
+
+1. Whenever "list" request returns a list of objects and a resource version "rv",
+starting a watch from the returned "rv" will never drop any events.
+2. For a given watch request (with resource version "rv"), the returned stream
+of events is always the same (e.g. very slow laggin watch may not cause dropped
+events). 
+
+ To satisfy the above requirements, we can't really rely only on the existing
+tools/frameworks that we have. Theoretically, we would be able to build the
+mapping from an object to list of nodes it is bound to using standard
+reflector/informer framework. But having separate informers for different kinds
+of objects (which is the only possible way to use them) may result in races.
+As an example, imagine a pod P referencing secret S, bounding pod P to node N
+happening right before creating secret S. Then the following race may potentially
+happen:
+1. user observes binding of pod P, he tries to retrieve the current value of a
+secret and it still doesn't exist. From now she knows that as soon as it will
+be created, it should be delivered via watch to him.
+2. however, pod informer is lagging in our code, and we first observe secret
+creation. It's not referenced yet (we didn't yet observe the pod), so we don't
+send any event.
+3. we observe the pod binding, but according to our semantics, we don't send
+any event.
+And we didn't send the expected "add secret" event to the watcher.
+
+ To solve the problem reliably, we need to be able to correctly serialize the
+watch events between different object types in a deterministic way.
+
+ One potential solution would be to identify this watch by a resource version
+combined from resource versions coming from different object kinds (e.g.
+pods have rv = rv1, secrets have rv = rv2, ...). Then we could keep the history
+of objects being processed to update the in-memory mapping and that would be
+deterministic. But the order might have been different e.g. after restarting
+apiserver. Which also means that it wouldn't work in HA setups (assuming we
+don't have some external "serializing" component).
+
+ Fortunately, we can solve it in much simpler way, with one additional assumption:
+
+1. All object types necessary to determine the in-memory mapping share the same
```

Comment:
I agree that this is limitation. Though it doesn't block sharding completely - it's still possible to shard a bunch of different resource types into separate etcd (e.g. Nodes, Services, Endpoints, ...). 

So maybe it's something we can leave with? [obviously, if you have any better suggestion how to solve it, I'm more than happy to hear it.]